modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/DistributedCacheQueryReducer.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.query;
+
+import java.util.UUID;
+import org.apache.ignite.internal.IgniteInterruptedCheckedException;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Reducer for distributed cache query.
+ */
+public interface DistributedCacheQueryReducer<T> extends CacheQueryReducer<T> {
```

Comment:
Let's have separate classes/interfaces hierarchy for local and distributed queries.
Because they have different contracts for similar methods. E.g. onLastPage() and onPage(nodeId, true);
Having both methods looks ambiguous.

Reply:
Reworked methods within hierarchy, rename some methods. So, now it looks better. Don't think that separation is good idea, as `reducer(): CacheQueryReducer` used within GridCacheQueryFutureAdapter and shared between all types of queries.

Reply:
CacheQueryReducer.onPage requires nodeId as a param. 
Was it made just because it is required for DistributedCacheQueryReducer?
Why do we need DistributedCacheQueryReducer at all?